metric_name_builder.go

```diff
@@ -90,14 +89,14 @@ var perUnitMap = map[string]string{
 // See rules at https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels,
 // https://prometheus.io/docs/practices/naming/#metric-and-label-naming
 // and https://github.com/open-telemetry/opentelemetry-specification/blob/v1.38.0/specification/compatibility/prometheus_and_openmetrics.md#otlp-metric-points-to-prometheus.
-func BuildCompliantMetricName(metric pmetric.Metric, namespace string, addMetricSuffixes bool) string {
+func BuildCompliantMetricName(name, unit string, metricType MetricType, namespace string, addMetricSuffixes bool) string {
```

Comment:
How will you add new parameters/functionalities without making breaking changes?
Looking at parameters like `addMetricSuffixes` I would expect that you might want to add some new feautures like this in future.

Would all metric types have the same parameters? Maybe you should have a builder function (or even type) for each metric type?

Reply:
> How will you add new parameters/functionalities without making breaking changes?
Looking at parameters like addMetricSuffixes I would expect that you might want to add some new feautures like this in future.

Since the Prometheus naming convention is stable, I don't see us adding new functionalities in the future.

> Would all metric types have the same parameters? Maybe you should have a builder function (or even type) for each metric type?

While we both know that different metric types may receive different suffixes (or none), this might not be true for all library users. I believe we can abstract that from the users of our library. Just tell us your metric name and type, and we'll give you the correct output. I don't want people to guess which function they should call

Reply:
I was thinking of something like


```go
type NonMonotonicCounterNamer struct {
  // optional parameters can go here
  Namespace string 
  MetricSuffixes bool
}

// required params can go here
func (n *NamerNonMonotonicCounter) Name(name, unit string) string {
```

This way could be easier for maintaining backwards-compatibility as well having different options for different metric types. However, if you are sure that it is unnecessary then it does not make sense. As this approach involves a lot of additional API surface (new types, fields)

Reply:
In the Otel/Prom meeting we thought that variadic arguments are not ideal for a hot-path function if it's not explicitly needed. The other ideas -- single options struct, namer type, etc, all seem fine from a performance standpoint so let's go with whichever one fits best into how the code is being used.

Reply:
> In the Otel/Prom meeting we thought that variadic arguments are not ideal for a hot-path function

I agree. I am also not their fan :wink: